File: backend/gists_utils/gists_utils/mariadb_handler.py
from os import getenv
from datetime import timezone
import mariadb

from gists_utils.types import FeedInfo, Gist, SearchResult
from gists_utils.logger import get_logger


class MariaDbHandler:
    def __init__(self):
        self._host = getenv("DB_HOSTNAME")
        self._db_name = "thegistofitsec"
        self._username = getenv("DB_USERNAME")
        self._password = getenv("DB_PASSWORD")
        self._logger = get_logger("mariadb_handler")
        self._connection = self._connect()

    def _connect(self) -> mariadb.Connection:
        try:
            return mariadb.connect(
                user=self._username,
                password=self._password,
                host=self._host,
                database=self._db_name,
                autocommit=True
            )
        except mariadb.Error as e:
            self._logger.error("Error connecting to mariadb", exc_info=True)
            raise e
    
    @staticmethod
    def query_response_to_gist(response: tuple) -> Gist:
        (
            id, reference, feed_id, author, title, published, 
            updated, link, summary, tags, search_query, disabled
        ) = response
        return Gist(
            id, reference, feed_id, author, title, 
            published.replace(tzinfo=timezone.utc), 
            updated.replace(tzinfo=timezone.utc), 
            link, summary, tags.split(";;"), search_query
        )
        
    def get_gist_by_id(self, gist_id: int) -> Gist | None:
        query = "SELECT * FROM gists WHERE id = ? AND disabled = FALSE"
        try:
            with self._connection.cursor() as cur:
                cur.execute(query, (gist_id,))
                result = cur.fetchone()
                return None if result is None else self.query_response_to_gist(result)
        except mariadb.Error as e:
            self._logger.error(f"Error getting gist with id '{gist_id}'", exc_info=True)
            raise e
    
    def get_feed_by_id(self, gist_id: int) -> FeedInfo | None:
        query = "SELECT * FROM feeds WHERE id = ?"
        try:
            with self._connection.cursor() as cur:
                cur.execute(query, (gist_id,))
                results = cur.fetchall()
        except mariadb.Error as e:
            self._logger.error(f"Error getting feed with id {gist_id}", exc_info=True)
            raise e
        if len(results) > 1:
            self._logger.error(f"Found multiple feeds with id {gist_id}")
            raise KeyError(gist_id)
        if len(results) == 0:
            return None
        return FeedInfo(*results[0])
    
    @staticmethod
    def query_response_to_search_result(response: tuple) -> SearchResult:
        result_id, gist_id, title, snippet, link, display_link, thumbnail_link, image_link = response
        return SearchResult(result_id, gist_id, title, snippet, link, display_link, thumbnail_link, image_link)
    
    def get_search_results_by_gist_id(self, gist_id: int) -> list[SearchResult]:
        query = "SELECT * FROM search_results WHERE gist_id = ?"
        try:
            with self._connection.cursor() as cur:
                cur.execute(query, (gist_id,))
                return [self.query_response_to_search_result(response) for response in cur.fetchall()]
        except mariadb.Error as e:
            self._logger.error(f"Error getting search results for gist with id {gist_id}", exc_info=True)
            raise e

File: GistBackend/GistBackend/Handler/MariaDbHandler.cs
using Dapper;
using GistBackend.Exceptions;
using GistBackend.Types;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using MySqlConnector;

namespace GistBackend.Handler;

public record MariaDbHandlerOptions(
    string Server,
    string User,
    string Password,
    uint Port = 3306
) {
    public readonly string Database = "TheGistOfItSec";
};

public interface IMariaDbHandler {
    public Task<RssFeedInfo?> GetFeedInfoByRssUrlAsync(string rssUrl, CancellationToken ct);
    public Task<int> InsertFeedInfoAsync(RssFeedInfo feedInfo, CancellationToken ct);
    public Task UpdateFeedInfoAsync(RssFeedInfo feedInfo, CancellationToken ct);
    public Task<Gist?> GetGistByReferenceAsync(string reference, CancellationToken ct);
    public Task<int> InsertGistAsync(Gist gist, CancellationToken ct);
    public Task UpdateGistAsync(Gist gist, CancellationToken ct);
    public Task<GoogleSearchResult[]> GetSearchResultsByGistIdAsync(int gistId, CancellationToken ct);
    public Task InsertSearchResultsAsync(IEnumerable<GoogleSearchResult> searchResults, CancellationToken ct);
    public Task UpdateSearchResultsAsync(IEnumerable<GoogleSearchResult> searchResults, CancellationToken ct);
}

public class MariaDbHandler(IOptions<MariaDbHandlerOptions> options, ILogger<MariaDbHandler>? logger) : IMariaDbHandler {
    private readonly string _connectionString = new MySqlConnectionStringBuilder {
        Server = options.Value.Server,
        Port = options.Value.Port,
        Database = options.Value.Database,
        UserID = options.Value.User,
        Password = options.Value.Password
    }.ConnectionString;

    public async Task<RssFeedInfo?> GetFeedInfoByRssUrlAsync(string rssUrl, CancellationToken ct)
    {
        const string query = "SELECT Title, RssUrl, Language, Id FROM Feeds WHERE RssUrl = @RssUrl";
        var command = new CommandDefinition(query, new { RssUrl = rssUrl }, cancellationToken: ct);

        await using var connection = await GetOpenConnectionAsync(ct);

        var feedInfos = (await connection.QueryAsync<RssFeedInfo>(command)).ToArray();
        if (feedInfos.Length > 1) throw new Exception("Found multiple feeds in database");

        return feedInfos.FirstOrDefault();
    }

    public async Task<int> InsertFeedInfoAsync(RssFeedInfo feedInfo, CancellationToken ct)
    {
        const string query = """
            INSERT INTO Feeds (Title, RssUrl, Language)
                VALUES (@Title, @RssUrl, @Language);
            SELECT LAST_INSERT_ID();
        """;
        var command = new CommandDefinition(query, feedInfo, cancellationToken: ct);

        await using var connection = await GetOpenConnectionAsync(ct);
        try
        {
            return await connection.ExecuteScalarAsync<int>(command);
        }
        catch (MySqlException e)
        {
            logger?.LogError(e, "Inserting FeedInfo failed");
            throw;
        }
    }

    public async Task UpdateFeedInfoAsync(RssFeedInfo feedInfo, CancellationToken ct)
    {
        const string query = "UPDATE Feeds SET Title = @Title, Language = @Language WHERE RssUrl = @RssUrl";
        var command = new CommandDefinition(query, feedInfo, cancellationToken: ct);

        await using var connection = await GetOpenConnectionAsync(ct);
        try
        {
            var rowsAffected = await connection.ExecuteAsync(command);
            if (rowsAffected != 1) throw new DatabaseOperationException("Did not successfully update feed info");
        }
        catch (Exception e) when (e is MySqlException or DatabaseOperationException)
        {
            logger?.LogError(e, "Updating FeedInfo failed");
            throw;
        }
    }

    public async Task<Gist?> GetGistByReferenceAsync(string reference, CancellationToken ct)
    {
        const string query = """
            SELECT Reference, FeedId, Author, Title, Published, Updated, Url, Summary, Tags, SearchQuery, Id
                FROM Gists WHERE Reference = @Reference
        """;
        var command = new CommandDefinition(query, new { Reference = reference }, cancellationToken: ct);

        await using var connection = await GetOpenConnectionAsync(ct);

        var gists = (await connection.QueryAsync<Gist>(command)).ToArray();
        if (gists.Length > 1) throw new Exception("Found multiple gists in database");

        return gists.FirstOrDefault();
    }

    public async Task<int> InsertGistAsync(Gist gist, CancellationToken ct)
    {
        const string query = """
            INSERT INTO Gists
                (Reference, FeedId, Author, Title, Published, Updated, Url, Summary, Tags, SearchQuery)
                VALUES (
                    @Reference, @FeedId, @Author, @Title, @Published, @Updated, @Url, @Summary, @Tags, @SearchQuery
                );
            SELECT LAST_INSERT_ID();
        """;
        var command = new CommandDefinition(query, gist, cancellationToken: ct);

        await using var connection = await GetOpenConnectionAsync(ct);
        return await connection.ExecuteScalarAsync<int>(command);
    }

    public async Task UpdateGistAsync(Gist gist, CancellationToken ct)
    {
        const string query = """
            UPDATE Gists
                SET FeedId = @FeedId, Author = @Author, Title = @Title, Published = @Published, Updated = @Updated,
                    Url = @Url, Summary = @Summary, Tags = @Tags, SearchQuery = @SearchQuery
                WHERE Reference = @Reference;
        """;
        var command = new CommandDefinition(query, gist, cancellationToken: ct);

        await using var connection = await GetOpenConnectionAsync(ct);
        try
        {
            var rowsAffected = await connection.ExecuteAsync(command);
            if (rowsAffected != 1) throw new DatabaseOperationException("Did not successfully update gist");
        }
        catch (Exception e) when (e is MySqlException or DatabaseOperationException)
        {
            logger?.LogError(e, "Updating gist failed");
            throw;
        }
    }

    public async Task<GoogleSearchResult[]> GetSearchResultsByGistIdAsync(int gistId, CancellationToken ct)
    {
        const string query = """
            SELECT Title, Snippet, Url, DisplayUrl, ThumbnailUrl, ImageUrl, GistId FROM SearchResults
                WHERE GistId = @GistId;
        """;
        var command = new CommandDefinition(query, new { GistId = gistId }, cancellationToken: ct);

        await using var connection = await GetOpenConnectionAsync(ct);

        return (await connection.QueryAsync<GoogleSearchResult>(command)).ToArray();
    }

    public async Task InsertSearchResultsAsync(IEnumerable<GoogleSearchResult> searchResults, CancellationToken ct)
    {
        await using var connection = await GetOpenConnectionAsync(ct);
        await using var transaction = await connection.BeginTransactionAsync(ct);
        foreach (var searchResult in searchResults)
        {
            await InsertSearchResultAsync(searchResult, connection, transaction, ct);
        }
        await transaction.CommitAsync(ct);
    }

    public async Task UpdateSearchResultsAsync(IEnumerable<GoogleSearchResult> searchResults, CancellationToken ct)
    {
        await using var connection = await GetOpenConnectionAsync(ct);
        await using var transaction = await connection.BeginTransactionAsync(ct);
        var searchResultsArray = searchResults.ToArray();
        await DeleteSearchResultsForGistIdAsync(searchResultsArray.First().GistId, connection, transaction, ct);
        foreach (var searchResult in searchResultsArray)
        {
            await InsertSearchResultAsync(searchResult, connection, transaction, ct);
        }
        await transaction.CommitAsync(ct);
    }

    private static async Task InsertSearchResultAsync(GoogleSearchResult searchResult, MySqlConnection connection,
        MySqlTransaction transaction, CancellationToken ct)
    {
        const string query = """
            INSERT INTO SearchResults
                (GistId, Title, Snippet, Url, DisplayUrl, ThumbnailUrl, ImageUrl)
                VALUES (@GistId, @Title, @Snippet, @Url, @DisplayUrl, @ThumbnailUrl, @ImageUrl)
        """;
        var command = new CommandDefinition(query, searchResult, transaction, cancellationToken: ct);
        await connection.ExecuteAsync(command);
    }

    private static async Task DeleteSearchResultsForGistIdAsync(int gistId, MySqlConnection connection,
        MySqlTransaction transaction, CancellationToken ct)
    {
        const string query = "DELETE FROM SearchResults WHERE GistId = @GistId";
        var command = new CommandDefinition(query, new { GistId = gistId }, transaction, cancellationToken: ct);
        var rowsAffected = await connection.ExecuteAsync(command);
        if (rowsAffected == 0)
        {
            throw new DatabaseOperationException("Did not delete any search results");
        }
    }

    private async Task<MySqlConnection> GetOpenConnectionAsync(CancellationToken ct)
    {
        MySqlConnection? connection = null;
        try
        {
            connection = new MySqlConnection(_connectionString);
            await connection.OpenAsync(ct);
            return connection;
        }
        catch
        {
            if (connection is not null) await connection.DisposeAsync();
            throw;
        }
    }
}
